### PR TITLE
TC to validate backup and restore for a rancher project with multiple member 

### PR DESCRIPTION
**Pull request by sabrarhussaini**
Added a new TC **MultipleMemberProjectBackupAndRestoreForSingleNamespace** to verify the backup and restore for a rancher project with multiple members.
Below are the TC details:
Task:  _https://portworx.atlassian.net/browse/PB-5248_
Testrail: _https://portworx.testrail.net/index.php?/cases/view/84876_

Which issue(s) this PR fixes (optional)
Closes #PB-5248

Special notes for your reviewer:
Rancher-S3:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4185/
Rancher-NFS:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4187/

### CODE REVIEW

**Review by snigdha-px — CHANGES_REQUESTED** [review]

File: drivers/scheduler/rke/rke.go [hunk 1]
```diff
@@ -24,6 +25,7 @@ const (
 )
 
 var RancherMap = make(map[string]*RancherClusterParameters)
+var rancherUserPassword = "Password@123"
```
[the remark this answers is not in the document]
Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go:28]: Addressed this comment.

File: drivers/scheduler/rke/rke.go [hunk 2]
```diff
@@ -184,6 +186,99 @@ func (r *Rancher) GetProjectID(projectName string) (string, error) {
...
+			Labels:      userLabel,
+		}
+		newUser, err := r.client.User.Create(userRequest)
+		userIDList = append(userIDList, newUser.ID)
+		principalIDList = append(principalIDList, newUser.PrincipalIDs[0])
+		if err != nil {
+			return nil, fmt.Errorf("failed to create user: %w", err)
+		}
+		//adding a sleep of 1sec because almost all users are getting created in the same second
+		time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go:220]: The users are getting created very fast and there might be no need for them to be created in parallel. Kindly advise.

File: drivers/scheduler/rke/rke.go [hunk 3]
```diff
@@ -184,6 +186,99 @@ func (r *Rancher) GetProjectID(projectName string) (string, error) {
...
+			ProjectID:       projectId,
+			UserPrincipalID: principalID,
+			RoleTemplateID:  "project-member",
+		}
+		_, err := r.client.ProjectRoleTemplateBinding.Create(roleRequest)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return userIDList, nil
```
[the remark this answers is not in the document]
Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go:233]: Addressed the comment.

**Review by snigdha-px — CHANGES_REQUESTED** [review]

File: drivers/scheduler/rke/rke.go [hunk 4]
```diff
@@ -37,6 +39,18 @@ type RancherClusterParameters struct {
...
+	rand.Seed(time.Now().UnixNano())
+	const letters = "abcdefghijklmnopqrstuvwxyz"
+	randomBytes := make([]byte, length)
+	for i := range randomBytes {
+		randomBytes[i] = letters[rand.Intn(len(letters))]
+	}
+	randomString := string(randomBytes)
+	return randomString
+}
+
```
[the remark this answers is not in the document]
Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go]: Thanks for suggesting this, I have addressed the change in the recent commit.

**Review by snigdha-px — APPROVED** [review]
LGTM

**Review by snigdha-px — COMMENTED** [review]
@sabrarhussaini 
Please see why CBT checks are failing

File: drivers/scheduler/rke/rke.go [hunk 5]
```diff
@@ -184,6 +198,104 @@ func (r *Rancher) GetProjectID(projectName string) (string, error) {
 	return projectId, fmt.Errorf("no project matching the given projectName %s was found", projectName)
 }
 
+//CreateUsersForRancherProject Creates rancher users based on the supplied number and adds them to the project
```
Comment (mkoppal-px) [inline drivers/scheduler/rke/rke.go:201]: nit: add a space after `//`

File: drivers/scheduler/rke/rke.go [hunk 6]
```diff
@@ -184,6 +198,104 @@ func (r *Rancher) GetProjectID(projectName string) (string, error) {
...
+func (r *Rancher) CreateUsersForRancherProject(projectName string, numberOfUsers int) ([]string, error) {
```
Comment (mkoppal-px) [inline drivers/scheduler/rke/rke.go:202]: You can break this helper function down to smaller chunks.
First create a simple function `CreateUserForRancherProject(username string, project string)`
Then create a function `CreateMultipleUsersForRancherProject((projectName string, userIDList []string)` and call `CreateUserForRancherProject` inside it.

We should not have `RandomString` func in `rke.go`. If you want to create random users. Then you build an array with random usernames in the test case itself and pass that array to the function `CreateMultipleUsersForRancherProject`. Or loop the array and call `CreateUserForRancherProject` in the test case

Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go]: Thank you for the comment @mkoppal-px , I have addressed this comment in the latest commit. Please review.

**Review by vpinisetti-px — APPROVED** [review]
Except above minor, LGTM

File: drivers/scheduler/rke/rke.go [hunk 7]
```diff
@@ -184,6 +184,109 @@ func (r *Rancher) GetProjectID(projectName string) (string, error) {
...
+	userRequest := &rancherClient.User{
+		Username:    userName,
+		Password:    userPassword,
+		Name:        fmt.Sprintf("Test " + userName),
+		Annotations: userAnnotation,
+		Labels:      userLabel,
+	}
+	newUser, err := r.client.User.Create(userRequest)
+	if err != nil {
+		return "", fmt.Errorf("failed to create user: %w", err)
```
[the remark this answers is not in the document]
Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go]: Addressed the comment

File: drivers/scheduler/rke/rke.go [hunk 8]
```diff
@@ -184,6 +184,109 @@ func (r *Rancher) GetProjectID(projectName string) (string, error) {
 	return projectId, fmt.Errorf("no project matching the given projectName %s was found", projectName)
 }
 
+// CreateUserForRancherProject Creates a rancher user and adds the user to the project
+func (r *Rancher) CreateUserForRancherProject(projectName string, userName string, userPassword string) (string, error) {
```
Comment (mkoppal-px) [inline drivers/scheduler/rke/rke.go:188]: nit: You can call it `username` and `password`

**Review by mkoppal-px — APPROVED** [review]
LGTM

File: drivers/scheduler/rke/rke.go (hunk 8 above)
Comment (sabrarhussaini) [reply drivers/scheduler/rke/rke.go]: Addressed it.

**Pull request merged**